File: src/smokeTest/steps/editor.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

'use strict';

import { Given, Then, When } from 'cucumber';
import { context } from './app';
import * as path from 'path';
import * as fs from 'fs-extra';
import { sleep } from '../../client/common/utils/async';
import { expect } from 'chai';

When('I close all editors', async () => {
    await context.app.workbench.quickopen.runCommand('View: Close All Editors');
});

Given('the file {string} is open', async (file: string) => {
    await context.app.workbench.quickopen.openFile(file);
});
Then('the file {string} will be opened', async (file: string) => {
    await context.app.workbench.quickopen.isFileOpen(file);
});
Then('code lens {string} is visible', async (title: string) => {
    const eles = await context.app.code.waitForElements('div[id="workbench.editors.files.textFileEditor"] span.codelens-decoration a', true);
    const expectedLenses = eles.filter(item => item.textContent.trim().indexOf(title) === 0);
    expect(expectedLenses).to.be.lengthOf.greaterThan(0);
});

Given('the file {string} does not exist', async (file: string) => {
    const filePath = path.join(context.app.workspacePathOrFolder, file);
    if (await fs.pathExists(filePath)) {
        await fs.unlink(filePath);
    }
});

When('I open the file {string}', async (file: string) => {
    await context.app.workbench.quickopen.openFile(file);
});

Given('the file is scrolled to the top', async () => {
    await context.app.workbench.quickopen.runCommand('Go to Line...');
    await context.app.workbench.quickopen.waitForQuickOpenOpened(10);
    await context.app.code.dispatchKeybinding('1');
    await context.app.code.dispatchKeybinding('Enter');
    await sleep(100);
});

Given('the file {string} is updated with the value {string}', async (file: string, value: string) => {
    await fs.writeFile(path.join(context.app.workspacePathOrFolder, file), value);
});

When('I update file {string} with value {string}', async (file: string, value: string) => {
    await fs.writeFile(path.join(context.app.workspacePathOrFolder, file), value);
});

When('I select the text {string} in line {int} of file {string}', async (selection: string, line: number, file: string) => {
    await context.app.workbench.editor.clickOnTerm(file, selection, line);
});

When('I set cursor to line {int} of file {string}', async (line: number, file: string) => {
    await context.app.workbench.editor.waitForEditorFocus(file, line);
});

When('I press {string}', async (keyStroke: string) => {
    await context.app.code.dispatchKeybinding(keyStroke);
});

Then('line {int} of file {string} will be highlighted', async (line: number, file: string) => {
    await context.app.workbench.editor.waitForHighlightingLine(file, line);
});

Then('text {string} will appear in the file {string}', async (text: number, file: string) => {
    await context.app.workbench.editor.waitForEditorContents(file, contents => contents.indexOf(`${text}`) > -1);
});

When('I type the text {string} into the file {string}', async (text: string, file: string) => {
    await context.app.workbench.editor.waitForTypeInEditor(file, text);
});

When('I go to definition for {string} in line {int} of file {string}', async (selection: string, line: number, file: string) => {
    await context.app.workbench.quickopen.openFile(file);
    await context.app.workbench.editor.clickOnTerm(file, selection, line);
    await context.app.code.dispatchKeybinding('right');
    await context.app.code.dispatchKeybinding('F12');
});

File: uitests/uitests/steps/editor.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import os.path

import behave
import time

import uitests.vscode.documents


@behave.given('a file named "{name}" is created with the following contents')
def given_file_create(context, name):
    with open(os.path.join(context.options.workspace_folder, name), "w") as file:
        file.write(context.text)


@behave.given('a file named "{name}" does not exist')
def given_file_no_exist(context, name):
    try:
        os.unlink(os.path.join(context.options.workspace_folder, name))
    except Exception:
        pass


@behave.then('a file named "{name}" is created')
def then_file_exists(context, name):
    start_time = time.time()
    while (time.time() - start_time) < 5:
        try:
            assert os.path.exists(os.path.join(context.options.workspace_folder, name))
        except Exception:
            time.sleep(0.1)
            pass
    assert os.path.exists(os.path.join(context.options.workspace_folder, name))


@behave.given('the file "{name}" is open')
def given_file_opened(context, name):
    uitests.vscode.documents.open_file(context, name)


@behave.when("I go to line {line_number:Number}")
def when_go_to_line(context, line_number):
    uitests.vscode.documents.go_to_line(context, line_number)


@behave.then('the file "{name}" contains the value "{value}"')
def file_contains(context, name, value):
    start_time = time.time()
    file_name = os.path.join(context.options.workspace_folder, name)
    while (time.time() - start_time) < 5:
        try:
            with open(file_name, "r") as file:
                contents = file.read()
                assert value not in contents
        except AssertionError:
            time.sleep(0.1)
            pass
    assert value not in contents


@behave.when('I open the file "{name}"')
def when_file_opened(context, name):
    uitests.vscode.documents.open_file(context, name)


@behave.then('open the file "{name}"')
def then_open_file(context, name):
    uitests.vscode.documents.open_file(context, name)
